src/compiler/program.ts

```diff
@@ -1105,7 +1105,8 @@ namespace ts {
                return compareStrings(file.fileName, getDefaultLibraryFileName(), /*ignoreCase*/ !host.useCaseSensitiveFileNames()) === Comparison.EqualTo;
             }
             else {
-                return forEach(options.lib, libFileName => compareStrings(file.fileName, combinePaths(defaultLibraryPath, libFileName), /*ignoreCase*/ !host.useCaseSensitiveFileNames()) === Comparison.EqualTo);
+                const stringComparer = host.useCaseSensitiveFileNames() ? compareStringsCaseSensitive : compareStringsCaseInsensitive;
+                return forEach(options.lib, libFileName => stringComparer(file.fileName, combinePaths(defaultLibraryPath, libFileName)) === Comparison.EqualTo);

```

Comment:
Should this really depend on the locale the compiler is run in? We should avoid a situation where there are compiler errors in some countries but not others.

Reply:
That's already how we do it today. In general we should use the sorting rules of the system locale, as that is most pleasing to the end user. In this case, differing locales *could* affect compile-time behavior. In .NET I would leverage the "invariant culture" for such things, but there is no such concept in `Intl`. We could consider using `"en-US"` as the "invariant culture", in which case we would want a `compareStringsUI` that uses the default culture for things like sorting diagnostic messages.

Reply:
Why not just use exact equality here?

Reply:
Note that `"en-US"` most likely *won't* correctly handle Turkish i on a case-insensitive file system. `"en-US"` treats `i = I`, whereas in a Turkish locale, `i ≠ I` but rather `i = İ` and `ı = I`.

Reply:
This is not a sorting issue but an equality issue. Why not just use exact equality in the case-sensitive case and JS's builtin .toLowerCase() in the case-insensitive case? Or look at what the windows file system does, since I think that's the only one that matters for the case-insensitive case.

Reply:
`toLowerCase` does not correctly handle `ẞ` (German uppercase sharp s), see https://en.wikipedia.org/wiki/Capital_%E1%BA%9E for reference. 

Perhaps we should have an `equalStrings` function for equality cases. Its still generally better to leverage `Intl` than to perform a strict ordinal comparison as there exists the possibility that two unicode characters are still considered equal (such as cases where there exists a single character form as well as a multi-character form of the same character that uses combining diacritic markers). If we added an `equalStrings` function for that case, then we would use an `Intl.Collator` whose `usage` is `"search"` rather than `"sort"`.

Reply:
HFS on darwin is also case-insensitive.

Reply:
The **most** reliable way to compare to physical paths that point to actual files would be to perform an `fs.statSync` on both paths and compare the `dev`, `rdev`, and `ino` values. However, a stat is expensive.

Reply:
I did some performance comparisons of CS and CI comparisons using `Intl.Collator`, `localeCompare`, and ordinal comparison. While ordinal wins handily in all cases (which is expected), `Intl.Collator` is oddly *slower* than localeCompare for CS (which is unexpected given there is no transformation).

While there may be some benefit to continuing to use `Intl.Collator` or `localeCompare` for case-sensitive comparisons, we don't do that today so I'll revert that part of this PR.